File: sliding_window/121_best_time_to_buy_and_sell_stock.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

/*
You are given an array prices where prices[i] is the price of a given stock on the ith day.

You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.
https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
O(n), O(n)
*/
class Solution {
  public:
    int maxProfit(vector<int>& prices) {
      if (prices.empty()) {
        return 0;
      }
      int max_profit = 0;
      int min_price = prices[0];
      for (int i = 1; i < prices.size(); i++) {
        min_price = min(prices[i], min_price);
        max_profit = max(prices[i] - min_price, max_profit);
      }
      return max_profit;
    }
};


int main() {
  vector<int> price{7, 1, 5, 3, 6, 4};
  int ans{Solution().maxProfit(price)};
  cout << "ans= " << ans << "\n";
  return 0;
}

File: sliding_window/121_best_time_to_buy_and_sell_stock.py
from typing import List

"""
You are given an array prices where prices[i] is the price of a given stock on the ith day.

You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.
https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
O(n), O(n)
"""
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0

        # initialize
        max_profit = 0
        min_price = prices[0]
        for price in prices[1:]:
            max_profit = max(price - min_price, max_profit)
            min_price = min(price, min_price)
        return max_profit


if __name__ == "__main__":
    prices = [7, 1, 5, 3, 6, 4]
    print(Solution().maxProfit(prices))
